openlibrary/solr/update_work.py

```diff
@@ -739,53 +743,65 @@ def add(name, value):
         def add_list(name, values):
             doc[name] = list(values)
 
+        borrowable_editions = set()
+        printdisabled_editions = set()
+        open_editions = set()
+        unclassified_editions = set()
+
+        language_editions = {}
+
         pub_goog = set()  # google
         pub_nongoog = set()
         nonpub_goog = set()
         nonpub_nongoog = set()
-
-        public_scan = False
+        printdisabled = set()
         all_collection = set()
+        public_scan = False
         lending_edition = None
         in_library_edition = None
         lending_ia_identifier = None
-        printdisabled = set()
+
         for e in editions:
             if 'ocaid' not in e:
                 continue
+
+            assert isinstance(e['ocaid'], str)
+            ocaid = e['ocaid'].strip()
+            collections = e.get('ia_collection', [])
+            all_collection.update(collections)
+
+            if 'inlibrary' in collections:
+                borrowable_editions.add(ocaid)
+            elif 'printdisabled' in collections:
+                printdisabled_editions.add(ocaid)
+            elif e.get('access_restricted_item', False) == "true":
+                unclassified_editions.add(ocaid)
+            else:
+                public_scan = True
+                open_editions.add(ocaid)
+
+            # Legacy
+            if 'printdisabled' in collections:
+                printdisabled.add(re_edition_key.match(e['key']).group(1))
+            # partners may still rely on these legacy fields, leave logic unchanged
             if not lending_edition and 'lendinglibrary' in e.get('ia_collection', []):
                 lending_edition = re_edition_key.match(e['key']).group(1)
                 lending_ia_identifier = e['ocaid']
             if not in_library_edition and 'inlibrary' in e.get('ia_collection', []):
                 in_library_edition = re_edition_key.match(e['key']).group(1)
                 lending_ia_identifier = e['ocaid']
-            if 'printdisabled' in e.get('ia_collection', []):
-                printdisabled.add(re_edition_key.match(e['key']).group(1))
-            all_collection.update(e.get('ia_collection', []))
-            assert isinstance(e['ocaid'], str)
-            i = e['ocaid'].strip()
-            if e.get('public_scan'):
-                public_scan = True
-                if i.endswith('goog'):
-                    pub_goog.add(i)
-                else:
-                    pub_nongoog.add(i)
-            else:
-                if i.endswith('goog'):
-                    nonpub_goog.add(i)
-                else:
-                    nonpub_nongoog.add(i)
+
         ia_list = (
-            list(pub_nongoog)
-            + list(pub_goog)
-            + list(nonpub_nongoog)
-            + list(nonpub_goog)
+            list(open_editions) +
```

Comment:
Previously we preferred non-google-based open IA ids, because the google based ones are low quality (eg https://archive.org/details/shakespearestra03gollgoog/page/n166/mode/2up ). Can we do that still? I think surfacing the goog scans more would result in a bad experience.

This might be a good place for `multisort_best`. So we'd sort first by availability (giving `0: open, 1: borrowable, 2: PD, 3: other`), then by ia_provider (giving probably `0: other, 1: goog books, 2: librivox books`), and then take the best. See https://github.com/internetarchive/openlibrary/blob/f74f7ab22661b6d920d6c7e4de320a99bbcc87fd/openlibrary/book_providers.py#L293-L299